src/net/neighbors/walk.rs

```diff
@@ -0,0 +1,2342 @@
+                // we know about this neighbor, but its key didn't match the
+                // neighboraddress.  Only try to re-connect with it if we haven't done
+                // so recently, so a rogue neighbor can't force us to DDoS another
+                // peer.
+                if n.last_contact_time + NEIGHBOR_MINIMUM_CONTACT_INTERVAL < get_epoch_time_secs() {
+                    to_resolve.push((*naddr).clone());
+                } else {
+                    // recently contacted
+                    resolved.insert(naddr.clone(), n);
+                }
+            } else {
+                // okay, we really don't know about this neighbor
+                to_resolve.push((*naddr).clone());
+            }
+        }
+        Ok((resolved, to_resolve))
+    }
+
+    /// Select neighbors that are routable, and ignore ones that are not.
+    /// TODO: expand if we ever want to filter by unroutable network class or something
+    fn filter_sensible_neighbors(neighbors: Vec<NeighborAddress>) -> Vec<NeighborAddress> {
+        let mut ret = vec![];
+        for neighbor in neighbors.into_iter() {
+            if neighbor.addrbytes.is_anynet() {
+                continue;
+            }
+            ret.push(neighbor);
+        }
+        ret
+    }
+
+    /// Given a neighbor we tried to insert into the peer database, find one of the existing
+    /// neighbors it collided with.  Return its slot in the peer db.
+    fn find_replaced_neighbor_slot(
+        conn: &DBConn,
+        nk: &NeighborKey,
+    ) -> Result<Option<u32>, net_error> {
+        let mut slots = PeerDB::peer_slots(conn, nk.network_id, &nk.addrbytes, nk.port)
+            .map_err(net_error::DBError)?;
+
+        if slots.len() == 0 {
+            // not present
+            return Ok(None);
+        }
+
+        let mut rng = thread_rng();
+        slots.shuffle(&mut rng);
+
+        for slot in slots {
+            let peer_opt =
+                PeerDB::get_peer_at(conn, nk.network_id, slot).map_err(net_error::DBError)?;
+
+            match peer_opt {
+                None => {
+                    continue;
+                }
+                Some(_) => {
+                    return Ok(Some(slot));
+                }
+            }
+        }
+
+        Ok(None)
+    }
+
+    /// Add a neighbor or schedule it to be pinged since it's up for replacement.
+    /// Returns (was-new?, neighbor)
+    fn add_or_schedule_replace_neighbor<'a>(
+        &mut self,
+        tx: &mut DBTx<'a>,
+        block_height: u64,
+        naddr: &NeighborAddress,
+        peer_version: u32,
+        network_id: u32,
+        handshake: &HandshakeData,
+        db_data: Option<&StackerDBHandshakeData>,
+    ) -> Result<(bool, Neighbor), net_error> {
+        let mut neighbor_from_handshake =
+            Neighbor::from_handshake(tx, peer_version, network_id, handshake)?;
+
+        if Neighbor::from_neighbor_address(tx, network_id, block_height, naddr)?.is_some() {
```

Comment:
In this PR, I'll `Neighbor::from_handshake()` to return `(Neighbor, bool)` to indicate whether or not the returned `Neighbor` represents DB-backed state.  This would obviate the call to `Neighbor::from_neighbor_address()`.

Reply:
I don't think the return types need to change at all here. Simple renaming would make it much more clear what these methods are doing. I'd suggest renaming `from_handshake()` to `Neighbor::load_and_update()` -- the "handshake" is implicit anyways because it accepts handshake data as a parameter, and what the method is doing is "loading" and "updating" (the method itself invokes a "get" and then "update" method).

And then I'd suggest renaming `from_neighbor_address()` to `load_by_address` or similar -- it's not "constructing a neighbor from an address" (which is what `from_neighbor_address` would imply), it's loading it from a db.

Reply:
Sure, I can do that